File: 36.有效的数独.py
#
# @lc app=leetcode.cn id=36 lang=python3
#
# [36] 有效的数独
#

# @lc code=start


from collections import defaultdict


class Solution:
    def isValidSudoku(self, board):
        row, col, block = defaultdict(set), defaultdict(set), defaultdict(set)
        for y in range(9):
            for x in range(9):
                val = board[y][x]
                if val == '.':
                    continue
                # 0,1,2
                # 3,4,5
                # 6,7,8
                # 计算处于哪个方格内，横坐标是x， 纵坐标是y
                point = (y // 3 * 3) + (x // 3)
                if val in row[y] or val in col[x] or val in block[point]:
                    return False
                row[y].add(val)
                col[x].add(val)
                block[point].add(val)
        return True


board = [["5", "3", ".", ".", "7", ".", ".", ".", "."],
         ["6", ".", ".", "1", "9", "5", ".", ".", "."],
         [".", "9", "8", ".", ".", ".", ".", "6", "."],
         ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
         ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
         ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
         [".", "6", ".", ".", ".", ".", "2", "8", "."],
         [".", ".", ".", "4", "1", "9", ".", ".", "5"],
         [".", ".", ".", ".", "8", ".", ".", "7", "9"],
         ]
s = Solution()
s.isValidSudoku(board)
# @lc code=end

File: 36.有效的数独.ts
/*
 * @lc app=leetcode.cn id=36 lang=typescript
 *
 * [36] 有效的数独
 */

// @lc code=start

// https://leetcode.cn/problems/valid-sudoku/solutions/46484/36-you-xiao-de-shu-du-by-alexer-660/
function isValidSudoku(board: string[][]): boolean {

  // 三个方向判重
  let rows = {};
  let columns = {};
  let boxes = {};
  // 遍历数独
  for (let i = 0; i < 9; i++) {
    for (let j = 0; j < 9; j++) {
      let num = board[i][j];
      if (num != '.') {
        // 子数独序号
        let boxIndex = Math.floor((i / 3)) * 3 + Math.floor(j / 3);
        if (rows[i + '-' + num] || columns[j + '-' + num] || boxes[boxIndex + '-' + num]) {
          return false;
        }
        // 以各自方向 + 不能出现重复的数字 组成唯一键值，若出现第二次，即为重复
        rows[i + '-' + num] = true;
        columns[j + '-' + num] = true;
        boxes[boxIndex + '-' + num] = true;
      }
    }
  }
  return true;

};
// @lc code=end

